### PR TITLE
Cold shard eviction

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:

Cold shard eviction. Keep filesystem usage under some limit. Do this by dropping least recently accessed shards as needed when downloading new shards. This is a general solution that works with random access.

- [x] Redesign shard states: `MISSING`, `DOWNLOADING`, `PRESENT` (evicting happens under the lock)
- [x] Initialize `shard_states` by walking and normalizing stream local dirs
- [x] Implement shard eviction
- [x] Redesign `StreamingDataset` keep-related args to add `cache_limit`, drop `keep_raw`
- [x] Bookkeeping for shard access times and tracking disk usage
- [x] Automatically evict shards based on disk usage limit

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -277,86 +276,80 @@ def __init__(self,
...
+
+            # Init each stream's local dir, discovering which shards are present.
+            are_shards_present = []
+            for stream in self.streams:
+                stream_shards = stream.get_shards(world)
+                are_shards_present += stream.init_local_dir(stream_shards)
+
+            # Use that to init shard states, shard access times, and cache usage.
+            self.cache_usage = 0
+            for shard_id, is_shard_present in enumerate(are_shards_present):
```
Comment (dakinggg) [inline streaming/base/dataset.py:331]: is this shard id consistent?

File: streaming/base/dataset.py [hunk 2]
```diff
@@ -366,21 +359,41 @@ def __del__(self) -> None:
...
+            int: Cache usage.
+        """
+        return int(self._cache_usage.get())
+
+    @cache_usage.setter
+    def cache_usage(self, cache_usage: int) -> None:
+        """Set the cache usage.
+
+        Args:
+            cache_usage (int): Cache usage.
```
Comment (dakinggg) [inline streaming/base/dataset.py:394]: what is cache usage? units?

File: streaming/base/dataset.py [hunk 3]
```diff
@@ -709,27 +834,22 @@ def __getitem__(self, sample_id: int) -> Any:
...
+            self._shard_access_times[shard_id] = time()
         except:
-            # Get handles to the shared shard states array and its protective file lock.
-            lock, shard_states = self._get_shard_states()
-
-            # Download the shard if not already being downloaded. Block if download in progress.
-            self._download_or_skip_shard(lock, shard_states, shard_id, True)
-
-            # Finally, access the sample.
+            # Proper path: shard wasn't present. Download the shard or wait for its download to
```
Comment (dakinggg) [inline streaming/base/dataset.py:846]: add to comment that download shard sets the access time, i was confused for a sec why the try sets the time and the except doesn't

File: streaming/base/format/base/reader.py [hunk 4]
```diff
@@ -63,6 +64,145 @@ def __len__(self) -> int:
         """
         return self.samples
 
+    def init_local_dir(self, ls: Set[str], keep_zip: bool) -> bool:
+        """Bring what shard files are present to a consistent state, returning whether prsent.
```
Comment (dakinggg) [inline streaming/base/format/base/reader.py:68]:
```suggestion
        """Bring what shard files are present to a consistent state, returning whether present.
```

File: streaming/base/format/base/reader.py [hunk 5]
```diff
@@ -63,6 +64,145 @@ def __len__(self) -> int:
...
+
+        Args:
+            ls (Set[str]): The listing of all files under dirname/[split/]. This is listed once
```
Comment (dakinggg) [inline streaming/base/format/base/reader.py:71]: more descriptive var name please

File: streaming/base/format/base/reader.py [hunk 6]
```diff
@@ -63,6 +64,145 @@ def __len__(self) -> int:
...
+
+    def get_zip_size(self) -> Optional[int]:
+        """Get the zip (compressed) size of this shard, if compression was used.
+
+        Returns:
+            Optional[int]: Size in bytes, or ``None`` if does not exist.
+        """
+        size = 0
+        for _, info in self.file_pairs:
+            if info is None:
```
Comment (dakinggg) [inline streaming/base/format/base/reader.py:164]: i guess it is guaranteed that if one `info` is `None`, all `info` are `None`?

File: streaming/base/stream.py [hunk 7]
```diff
@@ -173,8 +167,103 @@ def apply_default(self, default: Self) -> None:
...
+            if is_relative != has_proportion:
+                raise ValueError(f'Relative ("proportion") and absolute ("repeat", "samples", ' +
+                                 f'none) sub-dataset weights are incompatible with each other ' +
+                                 f'(error in stream {stream_id})')
+        return is_relative
+
+    @classmethod
+    def apply_weights(cls, streams: Sequence[Self], samples_per_stream: NDArray[np.int64],
+                      samples_per_epoch: Optional[int], seed: int) -> int:
+        """Given samples per stream, derive each stream's proportion/repeat/samples.
```
Comment (dakinggg) [inline streaming/base/stream.py:200]: note in docstring that this function modifies the streams objects to set the derived values

File: streaming/base/stream.py [hunk 8]
```diff
@@ -319,3 +407,27 @@ def get_shards(self, world: World) -> List[Reader]:
...
+        Args:
+            shards (List[Reader]): List of this stream's shards.
+
+        Returns:
+            List[bool]: List of whether each stream shard is present.
+        """
+        # List the cache directory (so that we hit the filesystem once).
+        local_dirname = os.path.join(self.local, self.split)
+        ls = set()
+        for dirname, _, subfiles in os.walk(local_dirname):
```
Comment (dakinggg) [inline streaming/base/stream.py:423]: are subdirs explicitly disallowed somewhere? otherwise this would just silently skip them, yeah?

File: streaming/base/shared/__init__.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2023 MosaicML Streaming authors
```
Comment (dakinggg) [inline streaming/base/shared/__init__.py:1]: didn't check the shared mem refactor closely, assuming its mostly copy pasted

**Review by dakinggg — COMMENTED** [review]
For testing, would be good to have
1. unit test for simple shard eviction (low cache limit)
2. mosaicgpt run with shard eviction to make sure it still runs, has good throughput, loss looks fine.

File: streaming/base/dataset.py (hunk 1 above)
Comment (knighton) [reply streaming/base/dataset.py:331]: yep -- (list of streams in provided order) -> (list of shard entries in provided order in the index.json for each stream)

File: streaming/base/dataset.py (hunk 2 above)
Comment (knighton) [reply streaming/base/dataset.py]: nice catch

File: streaming/base/stream.py (hunk 8 above)
Comment (knighton) [reply streaming/base/stream.py:423]: yeah we don't care about dirs, just shards

File: streaming/base/format/base/reader.py (hunk 6 above)
Comment (knighton) [reply streaming/base/format/base/reader.py:164]: yep, although not sure where or how to note this to the reader

File: streaming/base/format/base/reader.py (hunk 5 above)
Comment (knighton) [reply streaming/base/format/base/reader.py]: done

File: streaming/base/stream.py (hunk 7 above)
Comment (knighton) [reply streaming/base/stream.py:200]: done

File: streaming/base/dataset.py (hunk 3 above)
Comment (knighton) [reply streaming/base/dataset.py:846]: done

File: streaming/base/shared.py [hunk 10]
```diff
@@ -1,378 +0,0 @@
...
-        # Everybody waits until the flag is set to "go".
-        while not self.flag:
-            sleep(TICK)
-
-        # Note that we exited.
-        with self.lock:
-            self.num_exit += 1
-
-
-class CreateSharedMemory:
```
Comment (karan6181) [inline streaming/base/shared.py:169]: Why we removed `CreateSharedMemory` ? This class ensures proper cleanup and hides unwanted warning messages.

File: streaming/base/dataset.py [hunk 11]
```diff
@@ -659,79 +741,183 @@ def _get_work(self, world: World, epoch: int, sample_in_epoch: int) -> NDArray[n
...
+        This method is called internally by ``download_shard`` to clear space for more downloads.
+        """
+        # Get the shard with the oldest last access time.
+        shard_id = int(self._shard_access_times.numpy().argmin())
+        last_accessed = self._shard_access_times[shard_id]
+
+        # Verify the access time is valid (the shard has not been evicted). This will only pose a
+        # problem if there are no shards left to evict but you are trying to anyway.
+        if np.allclose(last_accessed, NEVER):
+            raise ValueError(f'Internal error: need to evict a shard to free up space, but no ' +
```
Comment (karan6181) [inline streaming/base/dataset.py:806]: Can we just throw a warning rather than an exception? What would be consequences if we throw a warning ?

File: streaming/base/dataset.py [hunk 12]
```diff
@@ -659,79 +741,183 @@ def _get_work(self, world: World, epoch: int, sample_in_epoch: int) -> NDArray[n
...
-        Returns:
-            Tuple[FileLock, NDArray[np.uint8]]: Lock, and array.
-        """
-        # Get the filelock that protects shard_states shared memory array.
-        lock = FileLock(self.shard_states_filename)
-
-        shard_states = np.ndarray(self.num_shards, buffer=self._shard_states.buf, dtype=np.uint8)
+            # Unknown state.
+            lock.release()
+            raise RuntimeError(f'Unknown shard state: {state}')
```
Comment (karan6181) [inline streaming/base/dataset.py:917]: More descriptive error message if possible ?

File: streaming/base/dataset.py [hunk 13]
```diff
@@ -747,138 +933,172 @@ def get_item(self, sample_id: int) -> Any:
...
         Returns:
-            Iterator[int]: Each sample ID, having been downloaded.
+            Iterator[int]: Each sample, having been downloaded.
         """
-        self._partition_state = _PartitionState(sample_ids)
-        Thread(target=self._download_thread, args=(self._partition_state,), daemon=True).start()
-        Thread(target=self._ready_thread, args=(self._partition_state,), daemon=True).start()
-        yield from self._partition_state
+        # Yield loop.
+        while True:
```
Comment (karan6181) [inline streaming/base/dataset.py:1062]: I get scared when I see infinite loop. Can we have some time to track and kill if it expires ?

File: streaming/base/format/base/reader.py [hunk 14]
```diff
@@ -74,6 +75,152 @@ def __len__(self) -> int:
         """
         return self.samples
 
+    def _evict_raw(self) -> None:
+        """Remove all raw files belonging to this shard."""
+        for raw_info, _ in self.file_pairs:
+            filename = os.path.join(self.dirname, self.split, raw_info.basename)
+            if os.path.exists(filename):
+                os.remove(filename)
```
Comment (karan6181) [inline streaming/base/format/base/reader.py:83]: Will there be a case where more than one thread/process is trying to delete a file? could be a edge case of race condition here where they pass line 82 !!! Thinking to either wrap `os.remove` in a `try..except` block where we ignore exception in `except` part or use `shutil.rmtree(filename, ignore_errors=True)` !

File: streaming/base/format/base/reader.py [hunk 15]
```diff
@@ -74,6 +75,152 @@ def __len__(self) -> int:
...
+
+    def _evict_zip(self) -> None:
+        """Remove all zip files belonging to this shard."""
+        for _, zip_info in self.file_pairs:
+            if zip_info:
+                filename = os.path.join(self.dirname, self.split, zip_info.basename)
+                if os.path.exists(filename):
+                    os.remove(filename)
```
Comment (karan6181) [inline streaming/base/format/base/reader.py:91]: same as above!

**Review by karan6181 — COMMENTED** [review]
What happens when `cache_limit` value is less than the individual shard file ? What is the behavior ?

File: streaming/base/dataset.py (hunk 11 above)
Comment (knighton) [reply streaming/base/dataset.py:806]: If we hit this problem, we can't evict a shard, which means we can't fulfill the purpose of this method

`download_shard()` needs this method to work to be able to have space to download a shard in all cases

I suppose we could return bool here instead of throwing an exception, and then throw an exception in the caller instead e.g. in download_shard, but that doesn't seem fundamentally better

File: streaming/base/format/base/reader.py (hunk 14 above)
Comment (knighton) [reply streaming/base/format/base/reader.py:83]: This is only called behind the lock `StreamingDataset._cache_filelock`

`StreamingDataset.evict_shard` -> takes lock -> `StreamingDataset._evict_shard` -> `Reader.evict` -> `Reader._evict_raw`

`StreamingDataset.download_shard` -> takes lock -> `StreamingDataset._evict_coldest_shard` -> `StreamingDataset._evict_shard` -> `Reader.evict` -> `Reader._evict_raw`

File: streaming/base/dataset.py (hunk 13 above)
Comment (knighton) [reply streaming/base/dataset.py:1062]: This loop is the epoch, no way to limit this

File: streaming/base/dataset.py [hunk 16]
```diff
@@ -19,70 +19,101 @@
...
-                if sample_id != -1:  # If -1, we skip.
-                    yield sample_id
-                self.yield_index += 1
-                continue
-            if self.is_stopped:
-                break
+        # Signal threads to exit.
+        with self._lock:
+            if self._is_exiting:
+                raise ValueError('Called exit() on an _IterState that is already exiting.')
```
[the remark this answers is not in the document]
Comment (knighton) [reply streaming/base/dataset.py]: Fixed

**Comment by knighton** [comment]
> What happens when `cache_limit` value is less than the individual shard file size? What is the behavior ?

throws exception

File: streaming/base/shared/prefix.py [hunk 17]
```diff
@@ -54,48 +62,87 @@ def _unpack_locals(data: bytes) -> Set[str]:
...
+    local working directory overlap. When attaching to an existing shm fails, we have reached the
+    end of the existing shms. We will register the next one.
+    """
+    prefix_int = 0
+    for prefix_int in _each_prefix_int():
+        prefix = f'{prefix_int:06}'
+        name = f'{prefix}_locals'
+        try:
+            shm = SharedMemory(name, False)
+        except:
```
Comment (karan6181) [inline streaming/base/shared/prefix.py:95]: any specific exception you want to catch such as `FileNotFoundError` or `FileExistError` ?

File: streaming/base/shared/prefix.py [hunk 18]
```diff
@@ -54,48 +62,87 @@ def _unpack_locals(data: bytes) -> Set[str]:
...
+    raise errs[-1]
+
+
+def get_shm_prefix(my_locals: List[str], world: World, retry: int = 7) -> Tuple[str, SharedMemory]:
+    """Register or lookup our shared memory prefix.
+
+    Args:
+        my_locals (List[str]): Local working dir of each stream, relative to /. We need to verify
+            that there is no overlap with any other currently running StreamingDataset.
+        world (World): Information about nodes, ranks, and workers.
```
Comment (karan6181) [inline streaming/base/shared/prefix.py:130]: nit: missing doc for `retry`.

**Comment by karan6181** [comment]
> > What happens when `cache_limit` value is less than the individual shard file size? What is the behavior ?
> 
> throws exception

Can you please add a test for the same ? Also worth to add a test when `cache_limit` is zero or negative value!

File: streaming/base/shared/prefix.py (hunk 17 above)
Comment (knighton) [reply streaming/base/shared/prefix.py]: done

File: streaming/base/shared/prefix.py (hunk 18 above)
Comment (knighton) [reply streaming/base/shared/prefix.py:143]: done

File: streaming/base/shared/prefix.py [hunk 19]
```diff
@@ -1,28 +1,36 @@
...
+"""Register or look up the prefix to use for all shared resources.
 
-For when using `threading` or `multiprocessing` from the python standard library won't do, because
-we are coordinating separately instantiated pytorch worker processes.
+The prefix is used by all workers using this StreamingDataset of this training job. This is used to
+prevent shared resources like shared memory and filelocks from colliding.
 """
 
 from multiprocessing import resource_tracker  # pyright: ignore
 from multiprocessing.shared_memory import SharedMemory
```
Comment (karan6181) [inline streaming/base/shared/prefix.py:11]: Duplicate import of `from multiprocessing.shared_memory import SharedMemory` since we already have `from streaming.base.shared import SharedMemory`.

File: tests/test_eviction.py [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+            assert set(os.listdir(local)) == set(
+                map(lambda f: f.replace('.zstd', ''), os.listdir(remote)))
+        else:
+            # Local has raw, remote has raw.
+            assert set(os.listdir(local)) == set(os.listdir(remote))
+
+    rmtree(local)
+
+
+def two(remote: str, local: str, keep_zip: bool):
```
Comment (dakinggg) [inline tests/test_eviction.py:44]: should this test check that all shards are present?

File: tests/test_eviction.py [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+                               local=local,
+                               keep_zip=keep_zip,
+                               cache_limit=1_000_000)
+    for _ in range(3):
+        for sample in dataset:  # pyright: ignore
+            pass
+    rmtree(local)
+
+
+def three(remote: str, local: str, keep_zip: bool):
```
Comment (dakinggg) [inline tests/test_eviction.py:58]: should this test check that not all shards are present?

File: tests/test_eviction.py [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+
+    assert os.listdir(local) == ['index.json']
+
+    for sample_id in range(dataset.num_samples):
+        dataset[sample_id]
+
+    assert set(os.listdir(local)) == full
+
+
+def five(remote: str, local: str, keep_zip: bool):
```
Comment (dakinggg) [inline tests/test_eviction.py:91]: this test looks the same as 3? is it supposed to do something different?

File: streaming/base/shared/memory.py [hunk 23]
```diff
@@ -1,43 +1,35 @@
...
-    """Create a new Shared Memory block or attach to an existing shared memory block.
+class SharedMemory:
+    """Improved quiet implementation of shared memory.
 
     Args:
-        name (Optional[str], optional): A unique shared memory block name. Defaults to ``None``.
-        create (Optional[bool], optional): Creates a new shared memory block or attaches to an
-            existing shared memory block. Defaults to ``None``.
+        name (str, optional): A unique shared memory block name. Defaults to ``None``.
+        create (bool, optional): Creates a new shared memory block or attaches to an existin
```
Comment (dakinggg) [inline streaming/base/shared/memory.py:21]:
```suggestion
        create (bool, optional): Creates a new shared memory block or attaches to an existing
```

File: streaming/base/shared/prefix.py [hunk 24]
```diff
@@ -54,48 +62,101 @@ def _unpack_locals(data: bytes) -> Set[str]:
...
+    Returns:
+        Tuple[str, SharedMemory]: Shared memory prefix and object. The name is required to be very
+            short due to limitations of Python on Mac OSX.
+    """
+    # Check my locals for overlap.
+    my_locals_set = _check_self(my_locals)
+
+    # First, the local leader registers the first available shm prefix, recording its locals.
+    if world.is_local_leader:
+        prefix_int = _check_and_find_retrying(my_locals_set, retry)
```
Comment (dakinggg) [inline streaming/base/shared/prefix.py:155]: If `_check_and_find_retrying` errors, this return value will be the error message. Is that intended? Should there be some error handling here?

File: streaming/base/shared/prefix.py [hunk 25]
```diff
@@ -1,28 +1,36 @@
...
-# Time out to wait before raising exception
-TIMEOUT = 60
+def _each_prefix_int() -> Iterator[int]:
+    """Get each possible prefix int to check in order.
+
+    Returns:
+        Iterator[int]: Each prefix int.
+    """
+    prefix_int = 0
+    while True:
```
Comment (dakinggg) [inline streaming/base/shared/prefix.py:31]: previously this loop was `range(10**6)`, should we maintain that behavior? Or is it intended for this to be infinite?

File: streaming/base/dataset.py [hunk 26]
```diff
@@ -277,6 +323,14 @@ def __init__(self,
         self.stream_per_shard = np.array(stream_per_shard, np.int64)
         self.num_shards = len(self.shards)
 
+        # Check that cache limit is possible.
+        if self.cache_limit:
+            min_cache_usage = sum(map(lambda stream: stream.get_index_size(), streams))
```
Comment (dakinggg) [inline streaming/base/dataset.py:328]: Isn't the min actually this size + the size of one shard? I guess that is harder to figure out up front?

File: streaming/base/dataset.py [hunk 27]
```diff
@@ -659,79 +763,183 @@ def _get_work(self, world: World, epoch: int, sample_in_epoch: int) -> NDArray[n
...
+            # Get the stream and shard.
             stream_id = self.stream_per_shard[shard_id]
             stream = self.streams[stream_id]
             shard = self.shards[shard_id]
+
+            # If cache_limit is enabled, we first may have to make space for the new shard.
+            if self.cache_limit:
+                # Here, we give the shard a provisional access time so it doesn't get inadvertently
+                # evicted if we need to evict shard(s).
+                self._shard_access_times[shard_id] = time()
```
Comment (dakinggg) [inline streaming/base/dataset.py:902]: at this point, this shard is missing, so its download time should be never, right? This line is safe, just checking my understanding because I think it is unnecessary.

**Review by dakinggg — COMMENTED** [review]
Could we also add a unit test that tests that the _coldest_ shard is actually the one that gets evicted?

File: streaming/base/dataset.py (hunk 26 above)
Comment (dakinggg) [reply streaming/base/dataset.py:328]: nvm, found the exception that happens in this case

**Review by karan6181 — COMMENTED** [review]
Can you please also add a test (With and Without Shard eviction) using `torch.utils.data.DataLoader` by varying `num_workers` ?

File: streaming/base/shared/memory.py [hunk 28]
```diff
@@ -47,28 +39,29 @@ def __init__(self,
...
                     self.created_shms.append(shm)
                 except FileExistsError:
                     sleep(TICK)
                     resource_tracker.register = self.fix_register
                     # Attaches to an existing shared memory block
-                    shm = SharedMemory(name, False, size)
+                    shm = BuiltinSharedMemory(name, False, size)
                     self.opened_shms.append(shm)
             self.shm = shm
+            self.buf = shm.buf
```
Comment (karan6181) [inline streaming/base/shared/memory.py:64]: `SharedMemory` buf is of type `MemoryView` and it is not-pickable when using `Spawn` multiprocessing. Instead of saving `buf` as a class object, create a method which returns a `buf` object. Something like this

```
@property
def buf(self):
    return self.shm.buf
```

File: streaming/base/dataset.py [hunk 29]
```diff
@@ -294,40 +348,79 @@ def __init__(self,
...
+            # Also use shard presence to initialize the shard states array and last access times.
+            for shard_id, is_shard_present in enumerate(are_shards_present):
+                if is_shard_present:
+                    self._shard_states[shard_id] = _ShardState.PRESENT
+                    self._shard_access_times[shard_id] = time()
+                else:
+                    self._shard_states[shard_id] = _ShardState.MISSING
+                    self._shard_access_times[shard_id] = NEVER
+
+        self._shared_barrier(world.workers_per_node)
```
Comment (karan6181) [inline streaming/base/dataset.py:414]: I see a hang when running a model training on more than one GPU. However, if I replace this line of code with torch dist barrier code then my training works. Looks like  `_shared_barrier` is not sync processes.

File: streaming/base/shared/prefix.py (hunk 25 above)
Comment (knighton) [reply streaming/base/shared/prefix.py:31]: infinite

low single digits in practice

File: streaming/base/shared/prefix.py (hunk 19 above)
Comment (knighton) [reply streaming/base/shared/prefix.py]: fixed

File: streaming/base/shared/prefix.py (hunk 24 above)
Comment (knighton) [reply streaming/base/shared/prefix.py:155]: eh?

**Comment by vchiley** [comment]
Tried running this with the llm-foundry repo
It runs without issue. Using the 125M model, I do not see any difference.
Screenshot 2023-05-09 at 12 54 25 PM
Disk utilization dropped a bit
Screenshot 2023-05-09 at 12 55 15 PM
but the same thing happened when I updt composer (without installing this branch) so I'm not sure this is from this PR
Screenshot 2023-05-09 at 12 56 23 PM

**Comment by knighton** [comment]
@karan6181 @vchiley I have integrated Karan's changes, please try it again?

- StreamingDataset init: switch shared barrier -> torch dist barrier
- StreamingDataset: _cache_filelock -> torch.multiprocessingManager().Lock()
- Ditto for SharedBarrier
- Exception handling in __iter__ threads

`dataset = StreamingDataset(..., cache_limit=(bytes <  total dataset bytes))`

**Review by karan6181 — APPROVED** [review]
